Add more support for CRD validation

fanzhangio opened the pull request:
#128

Liujingfang1 commented on cmd/internal/codegen/parse/crd.go:
```diff
@@ -173,6 +173,24 @@ var primitiveTemplate = template.Must(template.New("map-template").Parse(
+    Enum: {{ .Enum }},
+    {{ end -}}
+    {{ if .MaxLength -}}
+    MaxLength: {{ .MaxLength }},
```
I took a quick look and noticed something. `MaxLength` is a `*int64` type. You need some function here to convert the value to the correct type.

fanzhangio replied:
Good catch~ Thank you.

fanzhangio replied:
In my e2e test, I found an issue in crd generation and evaluation, I am going to fix it in this PR.

pwittrock commented on pkg/gen/apis/doc.go:
```diff
@@ -29,14 +29,35 @@ const Categories = "// +kubebuilder:categories="
 const Maximum = "// +kubebuilder:validation:Maximum="
 
-// ExclusiveMaximum annotates a numeric go struct field for CRD validation
+// ExclusiveMaximum annotates a boolean go struct field for CRD validation
```
+1.  I believe this should be numeric.

fanzhangio replied:
Good catch!  exclusiveMaximum and exclusiveMinimum are boolean values but do annotate numeric instances. So in the comments here it should be `numeric`.

fanzhangio commented on cmd/internal/codegen/parse/util.go:
```diff
@@ -202,3 +203,25 @@ func (c Comments) getTags(name, sep string) []string {
+// parseByteValue returns the literal digital number values from a byte array
+func parseByteValue(b []byte) string {
+	res := ""
+	elem := fmt.Sprintf("%v", b)
```
Enum type is `[ ]JSON`, and the raw data field of `JSON` struct is byte slice. In order to generate valid formatted go struct for `zz_generated.kubebuilder.go`, the {{.Validation}} string parts for template should contain the `raw` data field, not formatted string from byte. This util function is used to parse the raw digital value.

For example :
```golang
Enum: []v1beta1.JSON{v1beta1.JSON{[]byte{34, 76, 105, 111, 110, 34}}, v1beta1.JSON{[]byte{34, 87, 111, 108, 102, 34}}, v1beta1.JSON{[]byte{34, 68, 114, 97, 103, 111, 110, 34}}},
```

anfernee replied:
@fanzhangio I think you do the this without calling `fmt.Printf` where you assume the fmt format is like `[n n n n]`, which is pretty unreliable and less efficient. Here you simply need to print out the number and connect them with commas.

fanzhangio replied:
OK, so 
```golang
	token := fmt.Sprintf("%v", b)
	elem := strings.Join(strings.Fields(token), ",")
```
And then trim the brackets embraced around of elem.

anfernee replied:
I mean if you want to print out numbers and join them with `,` you can do it without using `fmt`.

fanzhangio commented on cmd/internal/codegen/parse/crd.go:
```diff
@@ -402,10 +424,12 @@ func getValidation(comment string, props *v1beta1.JSONSchemaProps) {
+		value := strings.Split(parts[1], ",")
+		enums := []v1beta1.JSON{}
+		for _, s := range value {
+			enums = append(enums, v1beta1.JSON{[]byte(`"` + s + `"`)})
```
Good question ! 

First, `MarshalJSON()` and `json.Marshal()` will be recursively called on this particular enum `[ ]JSON` by `JSONSchemaProps`based on its schema, but the go can not guarantee that the byte slice holds a `valid` JSON, .For example, the string word will get failed in marshaling the first character. Thus, for the string elements, it needs to wrap the elements with double quotation mark to indicate this is a string.

Moreover, I back to read the [JSON Schema Validation](https://datatracker.ietf.org/doc/draft-handrews-json-schema-validation/?include_text=1) and noticed that`Enum` validation keyword annotates any type of instance actually, the enum list could be number, string, `null` or something else. So, I updated here with applying a `IsNumeric()`for processing JSON types (number or string). I updated the e2e tests too,

The crd.yaml is like this 
```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  creationTimestamp: null
  labels:
    api: ""
    kubebuilder.k8s.io: unknown
  name: houses.got.sample.kubernetes.io
spec:
  group: got.sample.kubernetes.io
  names:
    kind: House
    plural: houses
  scope: Namespaced
  validation:
    openAPIV3Schema:
      properties:
        apiVersion:
          type: string
        kind:
          type: string
        metadata:
          type: object
        spec:
          properties:
            alias:
              enum:
              - Lion
              - Wolf
              - Dragon
              type: string
            knights:
              items:
                maxItems: 500
                minItems: 1
                type: string
              type: array
            name:
              maxLength: 15
              minLength: 1
              type: string
            power:
              exclusiveMinimum: true
              maximum: 100
              type: int
            rank:
              enum:
              - 1
              - 2
              - 3
              type: int
            winner:
              type: bool
          type: object
        status:
          type: object
      type: object
  version: v1beta1
status:
  acceptedNames:
    kind: ""
    plural: ""
  conditions: null

```

Liujingfang1 replied:
Thank you for the explanation and adding more cases in the test.

fanzhangio commented on test.sh:
```diff
@@ -549,6 +564,9 @@ prepare_staging_dir
 build_kb
 
+prepare_testdir_under_gopath
+test_crd_validation
```
Sure. Done

Liujingfang1 commented on test.sh:
```diff
@@ -478,6 +472,27 @@ EOF
+  kubebuilder create resource --group got --version v1beta1 --kind House
+
+  # Update crd
+  sed -i -e "s|type HouseSpec struct {|type HouseSpec struct{\n\n    // +kubebuilder:validation:Maximum=100\n    // +kubebuilder:validation:ExclusiveMinimum=true\n    Power int \`json:\"power\"\`\n\n    // +kubebuilder:validation:MaxLength=15\n    // +kubebuilder:validation:MinLength=1\n    Name string \`json:\"name\"\`\n\n    // +kubebuilder:validation:MaxItems=500\n    // +kubebuilder:validatio ...
```
Yes, we can add it in a separate PR.

pwittrock commented:
Lets target this for this weeks release on Thursday.

pwittrock commented on pkg/gen/apis/doc.go:
```diff
@@ -40,3 +40,24 @@ const ExclusiveMinimum = "// +kubebuilder:validation:ExclusiveMinimum="
+const Enum = "// +kubebuilder:validation:Enum="
+
+// MaxLength annotates a numeric length of go struct field for CRD validation
+const MaxLength = "// +kubebuilder:validation:MaxLength="
```
> // MaxLength specifies the maximum length of a string field as defined by https://tools.ietf.org/html/draft-fge-json-schema-validation-00

pwittrock commented on pkg/gen/apis/doc.go:
```diff
@@ -40,3 +40,24 @@ const ExclusiveMinimum = "// +kubebuilder:validation:ExclusiveMinimum="
+const MaxLength = "// +kubebuilder:validation:MaxLength="
+
+// MinLength annotates a numeric length of go struct field for CRD validation
+const MinLength = "// +kubebuilder:validation:MinLength="
```
> // MinLength specifies the minimum length of a string field as defined by https://tools.ietf.org/html/draft-fge-json-schema-validation-00

pwittrock commented on pkg/gen/apis/doc.go:
```diff
@@ -40,3 +40,24 @@ const ExclusiveMinimum = "// +kubebuilder:validation:ExclusiveMinimum="
+const MinLength = "// +kubebuilder:validation:MinLength="
+
+// MaxItems annotates a numeric quantity of struct field for CRD validation
+const MaxItems = "// +kubebuilder:validation:MaxItems="
```
> // MaxItems specifies the maximum number of items an array or slice field may contain as defined by https://tools.ietf.org/html/draft-fge-json-schema-validation-00

pwittrock commented on pkg/gen/apis/doc.go:
```diff
@@ -40,3 +40,24 @@ const ExclusiveMinimum = "// +kubebuilder:validation:ExclusiveMinimum="
+const MaxItems = "// +kubebuilder:validation:MaxItems="
+
+// MinItems annotates a numeric quantity of struct field for CRD validation
+const MinItems = "// +kubebuilder:validation:MinItems="
```
> // MinItems specifies the minimum number of items an array or slice field may contain as defined by https://tools.ietf.org/html/draft-fge-json-schema-validation-00

pwittrock commented on pkg/gen/apis/doc.go:
```diff
@@ -40,3 +40,24 @@ const ExclusiveMinimum = "// +kubebuilder:validation:ExclusiveMinimum="
+// MinItems annotates a numeric quantity of struct field for CRD validation
+const MinItems = "// +kubebuilder:validation:MinItems="
+
+// UniqueItems annotates a boolean go struct field for CRD validation
```
> // UniqueItems specifies that all values in an array or slice must be unique

pwittrock commented on pkg/gen/apis/doc.go:
```diff
@@ -40,3 +40,24 @@ const ExclusiveMinimum = "// +kubebuilder:validation:ExclusiveMinimum="
 const Pattern = "// +kubebuilder:validation:Pattern="
+
+// Enum annotates an enumerable go struct field for CRD validation
+const Enum = "// +kubebuilder:validation:Enum="
```
// Enum specifies the valid values for a field

pwittrock commented on cmd/internal/codegen/parse/crd.go:
```diff
@@ -173,6 +174,24 @@ var primitiveTemplate = template.Must(template.New("map-template").Parse(
+    MinLength: getInt({{ .MinLength }}),
+    {{ end -}}
+    {{ if .MaxItems -}}
+    MaxItems: getInt({{ .MaxItems }}),
```
MaxItems, MinItems, UniqueItems don't apply to primitives.  They apply to slices.  We should throw an error if we see them on primitive elements.

pwittrock replied:
Likewise, we should probably throw errors if we see things that apply to primitive values on non-primitive fields.

pwittrock commented on cmd/internal/codegen/parse/crd.go:
```diff
@@ -173,6 +174,24 @@ var primitiveTemplate = template.Must(template.New("map-template").Parse(
     {{ if .Format -}}
     Format: "{{ .Format }}",
     {{ end -}}
+    {{ if .EnumValue -}}
```
Add a // TODO on the variable to type check that the enum values match the type of the field - e.g. if the field is numberic the values should parse to be numeric.

fanzhangio replied:
Good. I have added //TODO comment on `EnumValue` and also implemented a numeric / string type check for parsing enum elements in **getValidation()** func

pwittrock commented:
@fanzhangio Thanks for catching that bug in this PR.

I left a few comments.  It would be great as a follow up to get test coverage on the actual usage of the validation annotations by writing tests that try to write objects that won't pass validation, and verifying that the validation fails.

We can do this in a similar manner to this PR:
https://github.com/kubernetes-sigs/kubebuilder/pull/181

fanzhangio commented:
@pwittrock Thanks for your comments.. I have updated the PR with:
- Update the comments in `pkg/gen/apis/doc.go`
- Changed **parseArrayValidation()** by adding `MaxItems`, `MinItems`, `UniqueItems` annotations handling.
- Added //TODO comment to `EnumItemsValue` for type check. Implemented `string` and `numeric` type check for parsing enum values in **getValidation()**
- Removed format print for byte value in **parseByteValue()** utils @anfernee 
- Fixed issues in primitive and array JSONSchema validation

anfernee commented on cmd/internal/codegen/parse/crd.go:
```diff
@@ -253,11 +265,28 @@ func (b *APIs) parseMapValidation(t *types.Type, found sets.String, comments []s
 var arrayTemplate = template.Must(template.New("array-template").Parse(
 	`v1beta1.JSONSchemaProps{
     Type:                 "array",
+    {{ if .MaxItems -}}
```
does this work? these fields don't match the template underneath. is there some kind of unit test to test this?

anfernee commented on cmd/internal/codegen/parse/crd.go:
```diff
@@ -410,9 +453,38 @@ func getValidation(comment string, props *v1beta1.JSONSchemaProps) {
+				// TODO support more types check
+				switch props.Type {
+				case "int", "int64", "uint64":
+					if _, err := strconv.ParseInt(s, 0, 32); err != nil {
```
32=>64

anfernee commented on cmd/internal/codegen/parse/crd.go:
```diff
@@ -410,9 +453,38 @@ func getValidation(comment string, props *v1beta1.JSONSchemaProps) {
+					}
+					enums = append(enums, v1beta1.JSON{[]byte(fmt.Sprintf("%v", s))})
+				case "int32", "unit32":
+					if _, err := strconv.ParseInt(s, 0, 64); err != nil {
```
64=>32

anfernee commented on cmd/internal/codegen/parse/crd.go:
```diff
@@ -410,9 +453,38 @@ func getValidation(comment string, props *v1beta1.JSONSchemaProps) {
+						log.Fatalf("Invalid integer value [%v] for a field of integer32 type", s)
+					}
+					enums = append(enums, v1beta1.JSON{[]byte(fmt.Sprintf("%v", s))})
+				case "float", "float32":
```
float handled with float64?

Liujingfang1 commented on cmd/internal/codegen/parse/crd.go:
```diff
@@ -410,9 +444,38 @@ func getValidation(comment string, props *v1beta1.JSONSchemaProps) {
-		for i := range enums {
-			props.Enum = append(props.Enum, v1beta1.JSON{[]byte(enums[i])})
+		if props.Type != "array" {
+			value := strings.Split(parts[1], ",")
```
I'm wondering if we can separate this code block of parsing Enum to a different function.  Then the function `getValidation` will keep short and we can also test `Enum` alone.

fanzhangio replied:
Agree. I wanted to do ti as well. I have encapsulated it into util functions.

droot commented on test.sh:
```diff
@@ -501,6 +501,54 @@ status:
+        instance.Spec=HouseSpec{Power:89.5,Knights:[]string{"Jaime","Bronn","Gregor Clegane"}, Alias:"Lion", Name:"Lannister", Rank:1}
+  ' pkg/controller/house/controller_test.go
+}
+
```
Please delete this test, once you implement PR which tests by creating a test project which exercises all the crd validation use cases.

droot commented on cmd/internal/codegen/parse/crd.go:
```diff
@@ -410,9 +444,38 @@ func getValidation(comment string, props *v1beta1.JSONSchemaProps) {
+			props.Enum = enums
 		}
 	case "Format":
 		props.Format = parts[1]
```
All of these are perfect candidates for unit-test :) but that will require some refactoring of the code which is outside scope of this PR.

fanzhangio replied:
Yes. I have encapsulated the handling logic into an individual util functions for better iteration in the future.
